backend/server/routers/utility.py

```diff
@@ -39,27 +41,43 @@ def get_core_courses(program: str, specialisations: list[str]):
          , [])
 
 
-def get_course_object(code: str, progTime: ProgramTime) -> Course:
+def get_course_object(code: str, prog_time: ProgramTime, locked_offering: tuple[int, int] | None = None, mark: int | None = 100) -> Course:
     ''' 
     This return the Course object for the given course code.
     Note the difference between this and the get_course function in courses.py
     '''
+    if mark is None:
+        mark = 100
     from server.routers.courses import terms_offered
-    years = "+".join(str(year) for year in range(progTime.startTime[0], progTime.endTime[0] + 1))
+    years = "+".join(str(year) for year in range(prog_time.startTime[0], prog_time.endTime[0] + 1))
     terms_result = terms_offered(code, years)["terms"]
-
-    new_terms_offered = {}
+    terms_possible = {}
     for year, terms in terms_result.items():
-        new_terms_offered[int(year)] = list(map(lambda x: int(x[1]), terms))
+        terms_possible[int(year)] = [int(term[1]) for term in terms]
 
     try:
         return Course(
             code,
             CONDITIONS[code],
-            100,
+            mark,
             COURSES[code]["UOC"],
-            new_terms_offered,
+            terms_possible,
+            locked_offering
         )
     except KeyError:
-        raise Exception(f"Course {code} not found (most likely course is discontinued)")
+        raise Exception(f"Course {code} not found (most likely course is discontinuation)")
 
+def extract_user_from_planner_data(plannerData: PlannerData) -> User:
+    """
+    Extracts a user object from the plannerData
+    """
+    user = User()
+    user.program = plannerData.program
+    user.specialisations = plannerData.specialisations
+    for _, year in enumerate(list(plannerData.plan)):
+        for term in year:
+            cleaned_term = {}
+            for course_name, course_value in term.items():
+                cleaned_term[course_name] = (course_value[0], course_value[1]) if course_value else (COURSES[course_name]["UOC"], None)
+            user.add_courses(cleaned_term)
+    return user
```

Comment:
if possible, this would ideally just be a method on `PlannerData`

Reply:
update: not possible because cyclic imports :(

will just revert to this and then add to backlog to fix this

Reply:
update 2: nvm its fixable with the import inside the method :)